### PR TITLE
[portmidi] Updated portmidi package to use alternative code source

### LINKED ISSUE #11974: [portmidi] Old version doesn't work on Win64
**Describe the bug**
Originally found when trying to use this package for https://github.com/csound/csound, any time the library API is called we get access violations/segmentation faults. A simple example using their test programs also causes the failures. The version of the code in use is from 2010. There have been numerous commits to their branch but no official release since.

Is it possible to pull down the latest SVN repo of portmidi to use as the package base instead of the zip file that is really old?

**Environment**
- OS: Win 10 latest
- Compiler: VS latest June 2020

**To Reproduce**
1. ./vcpkg install portmidi:x64-windows
2. Compile sample program e.g. midiclock.c and compile using VS.
3. Run app

**Expected behavior**
Program should operate as expected, accessing midi device and doing whatever.

**Failure logs** 
Access violation occurs when accessing portmidi API

**Additional context**

### PR DESCRIPTION
**Pull request by stekyne**
**Describe the pull request**
Updates the portmidi package to use the latest source. As discussed in #11974. Fixes #11974

- What does your PR fix? Fixes #
The previous version of the package was using a zip release from sourceforge which is 10 years old. It doesn't actually work with x64 Windows. There are commits after this release in the sourceforge repo that resolve the issue. I have migrated the sourceforge repo to git (under the Csound org github account). I'll update this going forward.

- Which triplets are supported/not supported? Have you updated the CI baseline?
x86/x64 windows are now supported

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?

### CODE REVIEW

File: ports/portmidi/CONTROL [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Source: portmidi
-Version: 0.217.1-1
+Version: r234
```
Comment (NancyLi1013) [inline ports/portmidi/CONTROL:2]:
```suggestion
Version: 2020-06-21
```
If there is no released version for the port, we should name the version as a date that you accept the commit.
Please refer to the [doc ](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/control-files.md#version)to learn more about this.

File: ports/portmidi/portfile.cmake [hunk 2]
```diff
@@ -4,15 +4,12 @@ if(VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")
     message(FATAL_ERROR "WindowsStore not supported")
 endif()
```
Comment (NancyLi1013) [inline ports/portmidi/portfile.cmake:6]: `include(vcpkg_common_functions)` is no longer needed here. Could you please remove this line?

Comment (NancyLi1013) [reply ports/portmidi/portfile.cmake:6]:
```
if(VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")
    message(FATAL_ERROR "WindowsStore not supported")
endif()
```
These can be replaced by `vcpkg_fail_port_install(ON_TARGET "uwp")`.

File: ports/portmidi/portfile.cmake [hunk 3]
```diff
@@ -35,6 +32,7 @@ vcpkg_configure_cmake(
```
Comment (NancyLi1013) [inline ports/portmidi/portfile.cmake:32]:
```
        -DJAVA_INCLUDE_PATH=
        -DJAVA_INCLUDE_PATH2=
        -DJAVA_JVM_LIBRARY=
```
These options seem to be not used. Could you please remove them?

File: ports/portmidi/portfile.cmake [hunk 4]
```diff
@@ -35,6 +32,7 @@ vcpkg_configure_cmake(
 
 # Run cmake build step, nothing is installed on Windows
 vcpkg_build_cmake()
+vcpkg_copy_pdbs()
```
Comment (NancyLi1013) [inline ports/portmidi/portfile.cmake:36]: Could you please update the way to handle the copyright as
`file(INSTALL ${SOURCE_PATH}/license.txt DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)`?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/portmidi/portfile.cmake (hunk 3 above)
Comment (stekyne) [reply ports/portmidi/portfile.cmake:32]: The build seems to fail without specifying these unfortunately.

**Comment by stekyne** [comment]
Thanks for the feedback. I've made the changes suggested except for the empty options one, this seems to be required by the package.

File: ports/portmidi/portfile.cmake (hunk 3 above)
Comment (NancyLi1013) [reply ports/portmidi/portfile.cmake:28]: Thanks. These variables seem to be needed on Linux.

File: ports/portmidi/portfile.cmake [hunk 5]
```diff
@@ -27,14 +20,15 @@ file(WRITE "${SOURCE_PATH}/pm_common/CMakeLists.txt" "${PM_CMAKE}")
 # Run cmake configure step
 vcpkg_configure_cmake(
     SOURCE_PATH ${SOURCE_PATH}
-    OPTIONS
-        -DJAVA_INCLUDE_PATH=
-        -DJAVA_INCLUDE_PATH2=
+    OPTIONS 
+	  	-DJAVA_INCLUDE_PATH=
+    	-DJAVA_INCLUDE_PATH2=
```
Comment (NancyLi1013) [inline ports/portmidi/portfile.cmake:25]:
```suggestion
	  -DJAVA_INCLUDE_PATH=
    	  -DJAVA_INCLUDE_PATH2=
```

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
@stekyne 
Thanks for your quick update.
Does this port support Linux and osx? Are you clear about this?

**Comment by stekyne** [comment]
Portmidi is crossplatform so it should work yes. I haven't tested these specific package changes on OSX/Linux though. As far as I can tell, they're fairly platform agnostic?

**Review by stekyne — COMMENTED** [review]
Resolved changes requested

**Comment by NancyLi1013** [comment]
Since I noticed that `portmidi` cannot build successfully on Linux and osx in CI build process.
Would you like to look into this?
https://dev.azure.com/vcpkg/public/_build/results?buildId=39183&view=logs&j=1d4f6642-7a41-537c-ff05-cbbc9ba4f53d&t=f56c2f90-bc2c-538e-3a85-64e370fdc032

You can enter the above link and search `portmidi`, then you will see the build results. If you have no linux or osx machine, you can remove portmidi:x64-linux=fail and portmidi:x64-osx=fail from ci.baseline.txt. Try to rebuild this, you can get the specific errors.

**Comment by stekyne** [comment]
Ok I'll need to take a look. It's missing Java but can't see how to disable that with a single flag. I'll get back to this shortly.

**Comment by ras0219-msft** [comment]
Could we use https://sourceforge.net/code-snapshots/svn/p/po/portmedia/code/portmedia-code-r234.zip instead, potentially with a small patch checked into this repo?

We strongly prefer not pulling unofficial forks if possible.

**Comment by stekyne** [comment]
@ras0219-msft sure I can do that. Your link no longer works though. I see you can download a snapshot of the repo on the site directly but it seems to be an ephemeral thing? Is that used by other packages?

**Comment by NancyLi1013** [comment]
@stekyne 
Could you please resolve the conflicts?

**Comment by stekyne** [comment]
@NancyLi1013 I will but I need some clarification on my last question. Is it ok to move forward with using a non-official fork of the source or do I need to use an official zip? The link provided doesn't work after a time, so I'm not sure it's viable to use that method.

**Comment by NancyLi1013** [comment]
@stekyne 
Where do you get the latest source code for `portmidi`? 
We doesn't recommend to use unofficial forks. 
So could you please update it as the original site?

**Comment by NancyLi1013** [comment]
@stekyne 
Could you please resolve the conflicts?

**Comment by stekyne** [comment]
@NancyLi1013 Sorry for the delay. I managed (with help) to get the original author to release a new zip of the source. It's been updated and I think the CI is passing. Can this be merged?

**Comment by NancyLi1013** [comment]
@stekyne 
Could you please also bump the version in `CONTROL `file?

I noticed that it is set as fail on x64-linux and x64-osx in ci.baseline.txt. 
Could you please help check if it supports linux and osx?

**Comment by stekyne** [comment]
@NancyLi1013 can you point me to the CI failure? I saw all green ticks on the details panel.

**Comment by NancyLi1013** [comment]
@stekyne 

I mean the related triplets were set as fail in ci.baseline.txt. https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt#L1360-L1361
```
portmidi:x64-linux=fail
portmidi:x64-osx=fail
```
I also checked the results on CI and they skipped. I tried to install it on my machine and it didn't work.

So could you help if it doesn't support linux and osx? If not, we'd better add `Supports: !(linux|osx)` to CONTROL and remove these codes from ci.baseline.txt.

**Comment by BillyONeal** [comment]
@stekyne Can you confirm that you're happy with my changes?

File: ports/portmidi/portfile.cmake [hunk 6]
```diff
@@ -1,13 +1,15 @@
-vcpkg_fail_port_install(ON_TARGET "UWP")
+vcpkg_fail_port_install(ON_TARGET "linux" "osx" ON_ARCH "arm" ON_TARGET "uwp")
```
Comment (strega-nil) [inline ports/portmidi/portfile.cmake:1]:
```suggestion
vcpkg_fail_port_install(ON_TARGET "linux" "osx" "uwp" ON_ARCH "arm")
```

I think this is nicer, imo.

File: ports/portmidi/CONTROL [hunk 7]
```diff
@@ -1,5 +1,5 @@
 Source: portmidi
-Version: 0.217.1-4
+Version: 0.234
 Homepage: https://sourceforge.net/projects/portmedia/
 Description: Free, cross-platform, open-source I/O library for MIDI
-Supports: !uwp
\ No newline at end of file
+Supports: windows&!uwp&!arm
```
Comment (strega-nil) [inline ports/portmidi/CONTROL:5]: `portmidi` supports OSX and Linux; we just don't have it working. Therefore, it shouldn't have the supports expression exclude these platforms.

**Comment by stekyne** [comment]
Thanks for taking a look. I've been trying to get it building for osx and linux in the mean time since updating. The cmake in portmidi is a bit inflexible so im trying to find out the right solution. Either fix it upstream or make a quick patch. Basically just need to allow building without jni support.

File: ports/portmidi/CONTROL (hunk 7 above)
Comment (strega-nil) [reply ports/portmidi/CONTROL:5]: OTOH, Billy argues that that's not valid, so I'll take this back for now. @ras0219 we should discuss this at the PR meeting.

**Pull request merged**